crates/router/src/db/api_keys.rs

```diff
@@ -126,55 +126,254 @@ impl ApiKeyInterface for Store {
+            // matter to someone using limit/offset in `list_api_keys_by_merchant_id`
+            locked_api_keys.remove(pos);
+            Ok(true)
+        } else {
+            Ok(false)
+        }
     }
 
     async fn find_api_key_by_merchant_id_key_id_optional(
         &self,
-        _merchant_id: &str,
-        _key_id: &str,
+        merchant_id: &str,
+        key_id: &str,
     ) -> CustomResult<Option<storage::ApiKey>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        Ok(self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .find(|k| k.merchant_id == merchant_id && k.key_id == key_id)
+            .cloned())
     }
 
     async fn find_api_key_by_hash_optional(
         &self,
-        _hashed_api_key: storage::HashedApiKey,
+        hashed_api_key: storage::HashedApiKey,
     ) -> CustomResult<Option<storage::ApiKey>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        Ok(self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .find(|k| k.hashed_api_key == hashed_api_key)
+            .cloned())
     }
 
+    #[allow(clippy::unwrap_used)]
     async fn list_api_keys_by_merchant_id(
         &self,
-        _merchant_id: &str,
-        _limit: Option<i64>,
-        _offset: Option<i64>,
+        merchant_id: &str,
+        limit: Option<i64>,
+        offset: Option<i64>,
     ) -> CustomResult<Vec<storage::ApiKey>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        // first gather all the keys for the given merchant_id, then apply any limit or offset, if
+        // requested. This uses memory inefficiently if a limit/offset is requested, but should be
+        // sufficient for the mockdb usecase
+        let mut keys_for_merchant_id: Vec<storage::ApiKey> = self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .filter(|k| k.merchant_id == merchant_id)
+            .cloned()
+            .collect();
+
+        // mimic the SQL limit/offset behavior
+        if let Some(offset) = offset {
+            if offset < 0 {
+                Err(errors::StorageError::MockDbError)?;
+            }
+            let offset: usize = offset.try_into().unwrap();
+            if offset >= keys_for_merchant_id.len() {
+                keys_for_merchant_id = vec![];
+            } else {
+                keys_for_merchant_id = keys_for_merchant_id[offset..].to_vec();
+            }
+        }
+
+        if let Some(limit) = limit {
+            if limit < 0 {
+                Err(errors::StorageError::MockDbError)?;
+            }
+            let limit: usize = limit.try_into().unwrap();
+            keys_for_merchant_id.truncate(limit);
+        }
```

Comment:
The [`skip()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip) and [`take()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take) methods should help simplify this logic, except for the negative number checks and `i64` to `usize` conversion.

Also, could you please update this to return an error in case the `i64` to `usize` conversion fails (and remove the `#[allow(clippy::unwrap_used)]` annotation)? This way, we can ensure that the `MockDb` implementation never panics, and we can call panicking functions in tests as necessary.

Reply:
done. I wanted to do more with `skip` and `take` but If i wanted to use only one iterator i d have to call with both skip and take and use 0 as a default for skip and the vector len as the default for take, let me know if you would prefer that over what I did  here though im happy to change

Reply:
It should be fine to use one iterator, use both `skip()` and `take()`, with the defaults you mentioned. Please do make that change.